### ISSUE TITLE
Sparkle 1.14 and macOS Sierra

### ISSUE THREAD

**Issue by mdhari** [sparkle-project/Sparkle#1031]
**TL:DR**

**Do I need to upgrade all the tools to 1.15 (BinaryDelta, sign_update, Sparkle.framework) if I'm generating deltas on a macOS box running Sierra?**

So long story short, my build server melted down. I was generating deltas from a El Captain box and now the new box is running Sierra. Luckily I saved my DSA keys into a code repo.

All the scripts that run BinaryDelta and sign_update still work so I get the delta files generated. But it looks like the Sparkle framework is strangely starting to block the install with the following errors found in console (I get the generic error popup within the App):

MyApp	subsystem: com.apple.securityd, category: cssm, enable_level: 0, persist_level: 0, default_ttl: 0, info_ttl: 0, debug_ttl: 0, generate_symptoms: 0, enable_oversize: 0, privacy_setting: 2, enable_private_data: 0
MyApp	subsystem: com.apple.securityd, category: handleobj, enable_level: 0, persist_level: 0, default_ttl: 0, info_ttl: 0, debug_ttl: 0, generate_symptoms: 0, enable_oversize: 0, privacy_setting: 2, enable_private_data: 0
MyApp	subsystem: com.apple.securityd, category: MDS_Scan, enable_level: 0, persist_level: 0, default_ttl: 0, info_ttl: 0, debug_ttl: 0, generate_symptoms: 0, enable_oversize: 0, privacy_setting: 2, enable_private_data: 0
MyApp	subsystem: com.apple.securityd, category: MDS_API, enable_level: 0, persist_level: 0, default_ttl: 0, info_ttl: 0, debug_ttl: 0, generate_symptoms: 0, enable_oversize: 0, privacy_setting: 2, enable_private_data: 0
MyApp	subsystem: com.apple.securityd, category: SSclnt, enable_level: 0, persist_level: 0, default_ttl: 0, info_ttl: 0, debug_ttl: 0, generate_symptoms: 0, enable_oversize: 0, privacy_setting: 2, enable_private_data: 0
MyApp	subsystem: com.apple.securityd, category: MDS_IO, enable_level: 0, persist_level: 0, default_ttl: 0, info_ttl: 0, debug_ttl: 0, generate_symptoms: 0, enable_oversize: 0, privacy_setting: 2, enable_private_data: 0
MyApp	subsystem: com.apple.securityd, category: mdslock, enable_level: 0, persist_level: 0, default_ttl: 0, info_ttl: 0, debug_ttl: 0, generate_symptoms: 0, enable_oversize: 0, privacy_setting: 2, enable_private_data: 0
MyApp	subsystem: com.apple.securityd, category: security_exception, enable_level: 0, persist_level: 0, default_ttl: 0, info_ttl: 0, debug_ttl: 0, generate_symptoms: 0, enable_oversize: 0, privacy_setting: 2, enable_private_data: 0
MyApp	UNIX error exception: 17
MyApp	subsystem: com.apple.securityd, category: DBOpen, enable_level: 0, persist_level: 0, default_ttl: 0, info_ttl: 0, debug_ttl: 0, generate_symptoms: 0, enable_oversize: 0, privacy_setting: 2, enable_private_data: 0
MyApp	subsystem: com.apple.securityd, category: dbsession, enable_level: 0, persist_level: 0, default_ttl: 0, info_ttl: 0, debug_ttl: 0, generate_symptoms: 0, enable_oversize: 0, privacy_setting: 2, enable_private_data: 0
MyApp	subsystem: com.apple.securityd, category: atomicfile, enable_level: 0, persist_level: 0, default_ttl: 0, info_ttl: 0, debug_ttl: 0, generate_symptoms: 0, enable_oversize: 0, privacy_setting: 2, enable_private_data: 0
MyApp	0x610000068d00 opened /private/var/db/mds/system/mdsDirectory.db: 50744 bytes
MyApp	0x610000068d00 closed /private/var/db/mds/system/mdsDirectory.db
MyApp	UNIX error exception: 17
MyApp	subsystem: com.apple.securityd, category: sslAsn1, enable_level: 0, persist_level: 0, default_ttl: 0, info_ttl: 0, debug_ttl: 0, generate_symptoms: 0, enable_oversize: 0, privacy_setting: 2, enable_private_data: 0
MyApp	CSSM Exception: -2147415792 CSSMERR_CSP_INVALID_KEY
MyApp	subsystem: com.apple.securityd, category: SecImpExp, enable_level: 0, persist_level: 0, default_ttl: 0, info_ttl: 0, debug_ttl: 0, generate_symptoms: 0, enable_oversize: 0, privacy_setting: 2, enable_private_data: 0
MyApp	subsystem: com.apple.securityd, category: SecAccess, enable_level: 0, persist_level: 0, default_ttl: 0, info_ttl: 0, debug_ttl: 0, generate_symptoms: 0, enable_oversize: 0, privacy_setting: 2, enable_private_data: 0
MyApp	subsystem: com.apple.securityd, category: freeKey, enable_level: 0, persist_level: 0, default_ttl: 0, info_ttl: 0, debug_ttl: 0, generate_symptoms: 0, enable_oversize: 0, privacy_setting: 2, enable_private_data: 0
MyApp	UNIX error exception: 17
MyApp	UNIX error exception: 17
MyApp	subsystem: com.apple.securityd, category: SecAccessReference, enable_level: 0, persist_level: 0, default_ttl: 0, info_ttl: 0, debug_ttl: 0, generate_symptoms: 0, enable_oversize: 0, privacy_setting: 2, enable_private_data: 0
MyApp	subsystem: com.apple.securityd, category: callback, enable_level: 0, persist_level: 0, default_ttl: 0, info_ttl: 0, debug_ttl: 0, generate_symptoms: 0, enable_oversize: 0, privacy_setting: 2, enable_private_data: 0
MyApp	UNIX error exception: 17
MyApp	subsystem: com.apple.securityd, category: policy, enable_level: 0, persist_level: 0, default_ttl: 0, info_ttl: 0, debug_ttl: 0, generate_symptoms: 0, enable_oversize: 0, privacy_setting: 2, enable_private_data: 0
MyApp	subsystem: com.apple.securityd, category: dateparse, enable_level: 0, persist_level: 0, default_ttl: 0, info_ttl: 0, debug_ttl: 0, generate_symptoms: 0, enable_oversize: 0, privacy_setting: 2, enable_private_data: 0
MyApp	UNIX error exception: 17
MyApp	subsystem: com.apple.securityd, category: keyDate, enable_level: 0, persist_level: 0, default_ttl: 0, info_ttl: 0, debug_ttl: 0, generate_symptoms: 0, enable_oversize: 0, privacy_setting: 2, enable_private_data: 0
MyApp	subsystem: com.apple.securityd, category: SecError, enable_level: 0, persist_level: 0, default_ttl: 0, info_ttl: 0, debug_ttl: 0, generate_symptoms: 0, enable_oversize: 0, privacy_setting: 2, enable_private_data: 0
MyApp	 [leaf AnchorTrusted]
MyApp	subsystem: com.apple.securityd, category: trustSettings, enable_level: 0, persist_level: 0, default_ttl: 0, info_ttl: 0, debug_ttl: 0, generate_symptoms: 0, enable_oversize: 0, privacy_setting: 2, enable_private_data: 0
MyApp	subsystem: com.apple.securityd, category: pinningQA, enable_level: 0, persist_level: 0, default_ttl: 0, info_ttl: 0, debug_ttl: 0, generate_symptoms: 0, enable_oversize: 0, privacy_setting: 2, enable_private_data: 0
MyApp	could not enable test hierarchy: ApplePinningAllowTestCertsAppleSWUpdateSigning not true
MyApp	SecStaticCode: verification failed (trust result 5, error -2147409622)
MyApp	MacOS error: -2147409622
MyApp	MacOS error: -2147409622
MyApp	Sparkle: ===== MyApp =====
MyApp	Sparkle: The update archive has a valid DSA signature, but the app is also signed with Code Signing, which is corrupted: Error Domain=NSOSStatusErrorDomain Code=-2147409622 "(null)" UserInfo={SecCSArchitecture=x86_64}. The update will be rejected.
MyApp	inside abortUpdateWithError
MyApp	abortUpdateWithError:error SUSparkleErrorDomain:3001 errorReason: Optional("The update is improperly signed.") withURL

Here's the thing though, I can run the BinaryDelta apply manually and the app patches fine. I feel like there is application code blocking me. I noticed you guys said on the main page that you updated all the tools to be compatible with the 10.12 SDK.  Would you happen Is this one of the issues that you guys solved in 1.15+?

**Comment by pornel** [comment]
From the log it seems that Sparkle *was* able to successfully apply the delta. Since it proceeded to checking Apple Code Signing it means that the delta has applied successfully and passed all the verification checks (all file contents and all permission bits were matching).

The problem seems to be with Apple Code Signing.

**Comment by pornel** [comment]
BTW, this is not us: ApplePinningAllowTestCertsAppleSWUpdateSigning and it looks suspcious. Do you have any "test certs" that Xcode could have used?

**Comment by mdhari** [comment]
I'm not familiar with the key either O.o 

We use the testing feature of Xcode to run unit tests, but we leave it unsigned for that part. Maybe it's unrelated.

**Comment by zorgiepoo** [comment]
Just as a tidbit, if you verify with `codesign -vvv` pass `--deep` along too.

**Comment by mdhari** [comment]
It was the code signing that was the issue. Certificate settings were wrong
in the keychain. Thanks for the help.
